### PR TITLE
SO-5834: generic dependency model

### PR DESCRIPTION
**Pull request by cmark**

### CODE REVIEW

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/DependencyDocument.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Objects;
+
+import com.b2international.snowowl.core.ResourceURIWithQuery;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * @since 8.12.0
+ */
+public final class DependencyDocument implements Comparable<DependencyDocument> {
```
Comment (apeteri) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/DependencyDocument.java:30]: This class might need extra annotations or dummy setters for the derived fields if the "ignore unknown properties" option is not enabled.

Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/internal/DependencyDocument.java:30]: Tests are running just fine, so not sure about this.

**Review by apeteri — APPROVED** [review]
LGTM!

**Review by apeteri — APPROVED** [review]
👍

**Comment by apeteri** [comment]
Does `snow-owl-ext` need a separate PR or is everything handled in classes of this project?

**Comment by cmark** [comment]
> Does `snow-owl-ext` need a separate PR or is everything handled in classes of this project?

There is an open PR in that repo: https://github.com/b2ihealthcare/snow-owl-ext/pull/2681

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceCreateRequest.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+		
+	}
+
+	private void checkBranchPath(final RepositoryContext context, final boolean create) {
+		// If no branch is created, the branch should already exist
+		if (!create && !branchExists(branchPath, context)) {
+			throw new BadRequestException("Branch path '%s' should point to an existing branch if given.", branchPath);
+		}
+		
+		// If the branch should be created, it branch should not exist, however
```
Comment (AAAlinaaa) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceCreateRequest.java]: typo: it branch

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceUpdateRequest.java [hunk 3]
```diff
@@ -60,100 +85,118 @@ public final void setExtensionOf(ResourceURI extensionOf) {
...
 		return false;
 	}
 
+	private boolean updateOid(TransactionContext context, String oldOid, Builder updated) {
+		if (oid == null || oid.equals(oldOid)) {
+			return false;
+		}
+		
+		if (!oid.isBlank()) {
+			final boolean oidExist = ResourceRequests.prepareSearch()
```
Comment (AAAlinaaa) [inline core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceUpdateRequest.java:199]: Is an admin context not needed here?

**Review by AAAlinaaa — APPROVED** [review]
LGTM

File: core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceUpdateRequest.java (hunk 3 above)
Comment (cmark) [reply core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseTerminologyResourceUpdateRequest.java:199]: Yes and in other locations as well. (there is a Jira issue about it)

**Review by apeteri — APPROVED** [review]

**Pull request merged**